File: EPI/19.1/swift/19.1.swift
public enum Status {
    case Empty
    case Blocked
}

public typealias Coord = (row: Int, col: Int)

public func path(from start: Coord, to end: Coord, in maze: inout [[Status]]) -> [Coord]? {
    if start == end {
        // We have found the end vertex.
        return [start]
    }

    // Mark the start vertex as visited.
    maze[start.row][start.col] = .Blocked

    for neighbor in neighbors(start, in: maze) {
        if let p = path(from: neighbor, to: end, in: &maze) {
            // A recursive call has found the end vertex.
            return [start] + p
        }
    }

    // No path to the end vertex was found.
    return nil
}

private func neighbors(_ coord: Coord, in maze: [[Status]]) -> [Coord] {
    let (row, col) = coord

    let potentialNeighbors = [
        (row - 1, col),
        (row + 1, col),
        (row, col - 1),
        (row, col + 1),
    ]

    return potentialNeighbors.filter { (r, c) in
        r >= 0 && r < maze.count &&
        c >= 0 && c < maze[r].count &&
        maze[r][c] == .Empty
    }
}

File: EPI/19.1/typescript/19.1.ts
export interface Coord {
    row: number;
    col: number;
}

export enum Color {
    White,
    Black,
}

export function path(maze: Color[][], start: Coord, end: Coord): Coord[] | undefined {
    if (equals(start, end)) {
        return [start];
    }

    maze[start.row][start.col] = Color.Black;

    for (const neighbor of neighbors(maze, start)) {
        if (maze[neighbor.row][neighbor.col] !== Color.Black) {
            const p = path(maze, neighbor, end);
            if (p) {
                return [start, ...p];
            }
        }
    }

    return undefined;
}

const equals = (a: Coord, b: Coord) => a.row === b.row && a.col === b.col;

const neighbors = (maze: Color[][], {row: r, col: c}: Coord) =>
    [
        {row: r - 1, col: c},
        {row: r, col: c + 1},
        {row: r + 1, col: c},
        {row: r, col: c - 1},
    ]
    .filter(({row: nr, col: nc}) =>
        nr >= 0 && nr < maze.length &&
        nc >= 0 && nc < maze[nr].length
    );
